### ISSUE TITLE
我觉得产品做得很好，推理速度能再快就就完美了

### ISSUE THREAD

**Issue by gg22mm** [PaddlePaddle/PaddleSeg#2501]
我觉得产品做得很好，推理速度能再快就就完美了

我用模型：modnet-mobilenetv2.yml 做分割

cpu_threads = 10
device = 'cpu' 
enable_mkldnn=True
enable_auto_tune = False

推理速度都在： 1.8736560344696045   到2秒这样，能不能再快些， 如果能达到0.* 秒以下就完美了。

不知道有什么办法，用的模型应该是最小的了吧？

图片分类检测时间: 0.2739129066467285
图片分类清晰度: 0.24283409118652344
分割时间: 1.943784475326538
目标检测时间: 1.2342767715454102

总时间: 3.695807933807373
后面还会添加模型， 就越来越慢了

**Comment by Bobholamovic** [comment]
可以尝试使用warmup，并且一次推理一组而不是一张图像，这样可以进一步提速~

**Comment by gg22mm** [comment]
> 是

循环推理也是一样的，我是进行多次的，不是一次一次。 你可以试一下， 实战一下才知真理，我可以提供模型给测试~~

**Comment by Bobholamovic** [comment]
好的。

实际上推理时间与输入图像尺寸、设备CPU、内存等因素有很大的关系，在[这里](https://github.com/PaddlePaddle/PaddleSeg/blob/develop/test_tipc/docs/benchmark_train.md)我们统计了不同模型的benchmark信息，使用标准配置时，在我们的服务器上FPS>1的模型应该还是有很多的。

**Comment by gg22mm** [comment]
> 间与输入图像尺寸、设备CPU、内存等因素有很大的
嗯， 但是对比一下

图片分类检测时间: 0.2739129066467285
图片分类清晰度: 0.24283409118652344
分割时间: 1.943784475326538
目标检测时间: 1.2342767715454102

看看有没有优化空间

cd PaddleSeg/Matting
[我的模型下载在这里]()  ,很难商用

**Comment by Bobholamovic** [comment]
我尝试在本地跑了这个模型，使用Python部署，只跑一张图，尺寸为1024x2048，使用的Paddle Inference配置与你提供的相同：

```
cpu_threads = 10
device = 'cpu'
enable_mkldnn=True
enable_auto_tune = False
```

得到性能指标如下：

```
2022-09-01 11:24:16 [INFO]      ---------------------- Env info ----------------------
2022-09-01 11:24:16 [INFO]       OS_version: CentOS Linux 7
2022-09-01 11:24:16 [INFO]       CUDA_version: 10.2.89
2022-09-01 11:24:16 [INFO]       CUDNN_version: 7.6.5
2022-09-01 11:24:16 [INFO]       drivier_version: 440.64.00
2022-09-01 11:24:16 [INFO]      ---------------------- Paddle info ----------------------
2022-09-01 11:24:16 [INFO]       paddle_version: 2.3.1
2022-09-01 11:24:16 [INFO]       paddle_commit: 3cc6ae69ed93388b2648bcc819d593130dede752
2022-09-01 11:24:16 [INFO]       log_api_version: 1.0
2022-09-01 11:24:16 [INFO]      ----------------------- Conf info -----------------------
2022-09-01 11:24:16 [INFO]       runtime_device: cpu
2022-09-01 11:24:16 [INFO]       ir_optim: True
2022-09-01 11:24:16 [INFO]       enable_memory_optim: True
2022-09-01 11:24:16 [INFO]       enable_tensorrt: False
2022-09-01 11:24:16 [INFO]       enable_mkldnn: True
2022-09-01 11:24:16 [INFO]       cpu_math_library_num_threads: 10
2022-09-01 11:24:16 [INFO]      ----------------------- Model info ----------------------
2022-09-01 11:24:16 [INFO]       model_name: 
2022-09-01 11:24:16 [INFO]       precision: fp32
2022-09-01 11:24:16 [INFO]      ----------------------- Data info -----------------------
2022-09-01 11:24:16 [INFO]       batch_size: 1
2022-09-01 11:24:16 [INFO]       input_shape: dynamic
2022-09-01 11:24:16 [INFO]       data_num: 1
2022-09-01 11:24:16 [INFO]      ----------------------- Perf info -----------------------
2022-09-01 11:24:16 [INFO]       cpu_rss(MB): 2217.2422, gpu_rss(MB): 12.0, gpu_util: 0.0%
2022-09-01 11:24:16 [INFO]       total time spent(s): 0.4558
2022-09-01 11:24:16 [INFO]       preprocess_time(ms): 119.0355, inference_time(ms): 336.7395, postprocess_time(ms): 0.0317
```

为了帮助我们更好地分析优化策略，请提供如下信息：

1. 测试时使用的硬件配置，如CPU型号、CPU核数、内存大小等；
2. 使用的部署方式，如PaddleSeg提供的C++或Python部署方式；
3. 输入图像的尺寸；
4. 我注意到你还使用了分类、检测等模型，方便透露一下分割模型是如何与其它模型耦合的吗？

**Comment by gg22mm** [comment]
1、我是在这样的环境测试出的结果，很慢，怎么破？：
![3](https://user-images.githubusercontent.com/6648329/187827545-22f0b839-6923-4da7-aaf3-da9891176e86.png)
![44](https://user-images.githubusercontent.com/6648329/187827839-5543ab42-84c3-46ca-87cd-d3ebaecd1db5.png)

2、多个模型我是通过paddle-inference 自己套一起整合一起的如：
![66](https://user-images.githubusercontent.com/6648329/187828173-50d18c2a-9c75-4849-ae95-43b9b238c31d.png)
![5](https://user-images.githubusercontent.com/6648329/187828185-19b23926-7e1a-4462-af86-86e2407c078f.png)
我是web访问所以+了flask。现在我这边访问有点慢，没办法商用，我的配置其实也不差了，对于web服务器来说，我觉得paddle飞浆是非常不错了，就是能再快点就可以大战宏图了~~


-----------------分割运行国程-------------------
2022-09-01 11:30:27 [INFO]      Using MKLDNN
I0901 11:30:28.002110  5372 analysis_predictor.cc:937] MKLDNN is enabled
e[1me[35m--- Running analysis [ir_graph_build_pass]e[0m
e[1me[35m--- Running analysis [ir_graph_clean_pass]e[0m
e[1me[35m--- Running analysis [ir_analysis_pass]e[0m
e[32m--- Running IR pass [mkldnn_placement_pass]e[0m
e[32m--- Running IR pass [simplify_with_basic_ops_pass]e[0m
e[32m--- Running IR pass [layer_norm_fuse_pass]e[0m
e[37m---    Fused 0 subgraphs into layer_norm op.e[0m
e[32m--- Running IR pass [attention_lstm_fuse_pass]e[0m
e[32m--- Running IR pass [seqconv_eltadd_relu_fuse_pass]e[0m
e[32m--- Running IR pass [seqpool_cvm_concat_fuse_pass]e[0m
e[32m--- Running IR pass [mul_lstm_fuse_pass]e[0m
e[32m--- Running IR pass [fc_gru_fuse_pass]e[0m
e[37m---    fused 0 pairs of fc gru patternse[0m
e[32m--- Running IR pass [mul_gru_fuse_pass]e[0m
e[32m--- Running IR pass [seq_concat_fc_fuse_pass]e[0m
e[32m--- Running IR pass [gpu_cpu_squeeze2_matmul_fuse_pass]e[0m
e[32m--- Running IR pass [gpu_cpu_reshape2_matmul_fuse_pass]e[0m
e[32m--- Running IR pass [gpu_cpu_flatten2_matmul_fuse_pass]e[0m
e[32m--- Running IR pass [matmul_v2_scale_fuse_pass]e[0m
e[32m--- Running IR pass [gpu_cpu_map_matmul_v2_to_mul_pass]e[0m
e[32m--- Running IR pass [gpu_cpu_map_matmul_v2_to_matmul_pass]e[0m
e[32m--- Running IR pass [matmul_scale_fuse_pass]e[0m
e[32m--- Running IR pass [gpu_cpu_map_matmul_to_mul_pass]e[0m
e[32m--- Running IR pass [fc_fuse_pass]e[0m
e[32m--- Running IR pass [repeated_fc_relu_fuse_pass]e[0m
e[32m--- Running IR pass [squared_mat_sub_fuse_pass]e[0m
e[32m--- Running IR pass [conv_bn_fuse_pass]e[0m
I0901 11:30:28.265028  5372 fuse_pass_base.cc:57] ---  detected 36 subgraphs
e[32m--- Running IR pass [conv_eltwiseadd_bn_fuse_pass]e[0m
e[32m--- Running IR pass [conv_transpose_bn_fuse_pass]e[0m
e[32m--- Running IR pass [conv_transpose_eltwiseadd_bn_fuse_pass]e[0m
e[32m--- Running IR pass [is_test_pass]e[0m
e[32m--- Running IR pass [runtime_context_cache_pass]e[0m
e[32m--- Running IR pass [depthwise_conv_mkldnn_pass]e[0m
I0901 11:30:28.288017  5372 fuse_pass_base.cc:57] ---  detected 17 subgraphs
e[32m--- Running IR pass [conv_bn_fuse_pass]e[0m
I0901 11:30:28.316011  5372 fuse_pass_base.cc:57] ---  detected 17 subgraphs
e[32m--- Running IR pass [conv_eltwiseadd_bn_fuse_pass]e[0m
e[32m--- Running IR pass [conv_transpose_bn_fuse_pass]e[0m
e[32m--- Running IR pass [conv_transpose_eltwiseadd_bn_fuse_pass]e[0m
e[32m--- Running IR pass [conv_bias_mkldnn_fuse_pass]e[0m
I0901 11:30:28.355996  5372 fuse_pass_base.cc:57] ---  detected 17 subgraphs
e[32m--- Running IR pass [conv_transpose_bias_mkldnn_fuse_pass]e[0m
e[32m--- Running IR pass [conv_elementwise_add_mkldnn_fuse_pass]e[0m
e[37m---    Fused 0 projection conv (as y) + elementwise_add patternse[0m
e[37m---    Fused 0 conv (as x) + elementwise_add patternse[0m
e[37m---    Fused 10 conv (as y) + elementwise_add patternse[0m
I0901 11:30:28.451968  5372 fuse_pass_base.cc:57] ---  detected 10 subgraphs
e[32m--- Running IR pass [conv_concat_relu_mkldnn_fuse_pass]e[0m
e[32m--- Running IR pass [conv_relu_mkldnn_fuse_pass]e[0m
I0901 11:30:28.470963  5372 fuse_pass_base.cc:57] ---  detected 1 subgraphs
e[32m--- Running IR pass [conv_leaky_relu_mkldnn_fuse_pass]e[0m
e[32m--- Running IR pass [conv_relu6_mkldnn_fuse_pass]e[0m
I0901 11:30:28.503949  5372 fuse_pass_base.cc:57] ---  detected 36 subgraphs
e[32m--- Running IR pass [conv_swish_mkldnn_fuse_pass]e[0m
e[32m--- Running IR pass [conv_hard_swish_mkldnn_fuse_pass]e[0m
e[32m--- Running IR pass [conv_mish_mkldnn_fuse_pass]e[0m
e[32m--- Running IR pass [conv_hard_sigmoid_mkldnn_fuse_pass]e[0m
e[32m--- Running IR pass [conv_gelu_mkldnn_fuse_pass]e[0m
e[32m--- Running IR pass [scale_matmul_fuse_pass]e[0m
e[37m---    fused 0 scale with matmule[0m
e[32m--- Running IR pass [reshape_transpose_matmul_mkldnn_fuse_pass]e[0m
e[37m---    Fused 0 ReshapeTransposeMatmul patterns for matmul Ope[0m
e[37m---    Fused 0 ReshapeTransposeMatmul patterns for matmul Op with transpose's xshapee[0m
e[37m---    Fused 0 ReshapeTransposeMatmul patterns for matmul Op with reshape's xshapee[0m
e[37m---    Fused 0 ReshapeTransposeMatmul patterns for matmul Op with reshape's xshape with transpose's xshapee[0m
e[32m--- Running IR pass [reshape_transpose_matmul_v2_mkldnn_fuse_pass]e[0m
e[37m---    Fused 0 ReshapeTransposeMatmul patterns for matmul_v2 Ope[0m
e[37m---    Fused 0 ReshapeTransposeMatmul patterns for matmul_v2 Op with transpose's xshapee[0m
e[37m---    Fused 0 ReshapeTransposeMatmul patterns for matmul_v2 Op with reshape's xshapee[0m
e[37m---    Fused 0 ReshapeTransposeMatmul patterns for matmul_v2 Op with reshape's xshape with transpose's xshapee[0m
e[32m--- Running IR pass [matmul_transpose_reshape_fuse_pass]e[0m
e[37m---    Fused 0 MatmulTransposeReshape patterns for matmul Ope[0m
e[32m--- Running IR pass [matmul_v2_transpose_reshape_fuse_pass]e[0m
e[37m---    Fused 0 MatmulTransposeReshape patterns for matmul_v2 Ope[0m
e[32m--- Running IR pass [batch_norm_act_fuse_pass]e[0m
e[37m---    fused 0 batch norm with relu activatione[0m
e[32m--- Running IR pass [softplus_activation_mkldnn_fuse_pass]e[0m
e[37m---    fused 0 softplus with relu activatione[0m
e[37m---    fused 0 softplus with tanh activatione[0m
e[37m---    fused 0 softplus with leaky_relu activatione[0m
e[37m---    fused 0 softplus with swish activatione[0m
e[37m---    fused 0 softplus with hardswish activatione[0m
e[37m---    fused 0 softplus with sqrt activatione[0m
e[37m---    fused 0 softplus with abs activatione[0m
e[37m---    fused 0 softplus with clip activatione[0m
e[37m---    fused 0 softplus with gelu activatione[0m
e[37m---    fused 0 softplus with relu6 activatione[0m
e[37m---    fused 0 softplus with sigmoid activatione[0m
e[32m--- Running IR pass [elt_act_mkldnn_fuse_pass]e[0m
e[37m---    fused 0 elementwise_add with relu activatione[0m
e[37m---    fused 0 elementwise_add with tanh activatione[0m
e[37m---    fused 0 elementwise_add with leaky_relu activatione[0m
e[37m---    fused 0 elementwise_add with swish activatione[0m
e[37m---    fused 0 elementwise_add with hardswish activatione[0m
e[37m---    fused 0 elementwise_add with sqrt activatione[0m
e[37m---    fused 0 elementwise_add with abs activatione[0m
e[37m---    fused 0 elementwise_add with clip activatione[0m
e[37m---    fused 0 elementwise_add with gelu activatione[0m
e[37m---    fused 0 elementwise_add with relu6 activatione[0m
e[37m---    fused 0 elementwise_add with sigmoid activatione[0m
e[37m---    fused 0 elementwise_sub with relu activatione[0m
e[37m---    fused 0 elementwise_sub with tanh activatione[0m
e[37m---    fused 0 elementwise_sub with leaky_relu activatione[0m
e[37m---    fused 0 elementwise_sub with swish activatione[0m
e[37m---    fused 0 elementwise_sub with hardswish activatione[0m
e[37m---    fused 0 elementwise_sub with sqrt activatione[0m
e[37m---    fused 0 elementwise_sub with abs activatione[0m
e[37m---    fused 0 elementwise_sub with clip activatione[0m
e[37m---    fused 0 elementwise_sub with gelu activatione[0m
e[37m---    fused 0 elementwise_sub with relu6 activatione[0m
e[37m---    fused 0 elementwise_sub with sigmoid activatione[0m
e[37m---    fused 0 elementwise_mul with relu activatione[0m
e[37m---    fused 0 elementwise_mul with tanh activatione[0m
e[37m---    fused 0 elementwise_mul with leaky_relu activatione[0m
e[37m---    fused 0 elementwise_mul with swish activatione[0m
e[37m---    fused 0 elementwise_mul with hardswish activatione[0m
e[37m---    fused 0 elementwise_mul with sqrt activatione[0m
e[37m---    fused 0 elementwise_mul with abs activatione[0m
e[37m---    fused 0 elementwise_mul with clip activatione[0m
e[37m---    fused 0 elementwise_mul with gelu activatione[0m
e[37m---    fused 0 elementwise_mul with relu6 activatione[0m
e[37m---    fused 0 elementwise_mul with sigmoid activatione[0m
e[1me[35m--- Running analysis [ir_params_sync_among_devices_pass]e[0m
e[1me[35m--- Running analysis [adjust_cudnn_workspace_size_pass]e[0m
e[1me[35m--- Running analysis [inference_op_replace_pass]e[0m
e[1me[35m--- Running analysis [ir_graph_to_program_pass]e[0m
I0901 11:30:28.774864  5372 analysis_predictor.cc:1007] ======= optimize end =======
I0901 11:30:28.775862  5372 naive_executor.cc:102] ---  skip [feed], feed -> img
I0901 11:30:28.782862  5372 naive_executor.cc:102] ---  skip [sigmoid_2.tmp_0], fetch -> fetch
100%|█████████████████████████████████████████████████████████100%|████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.02s/it]
2022-09-01 11:30:29 [INFO]      æ£å¾å®æ
################ 图像分割 end ####################### 分割时间: 1.8813960552215576

**Comment by Bobholamovic** [comment]
硬件配置还是相当不错的！不过代码上可能需要修改一下：由于初始化模型需要时间，因此我想应该预先初始化Predictor，在用户上传图片时仅调用已经构建好的Predictor的`predict()`方法，得到结果。

**Comment by gg22mm** [comment]
> 硬件配置还是相当不错的！不过代码上可能需要修改一下：由于初始化模型需要时间，因此我想应该预先初始化Predictor，在用户上传图片时仅调用已经构建好的Predictor的`predict()`方法，得到结果。

非常感谢！~~，这样做的话确实很快！ 多个模型下来才用了  总时间: 0.9087085723876953

**Comment by di-dii** [comment]
> 硬件配置还是相当不错的！不过代码上可能需要修改一下：由于初始化模型需要时间，因此我想应该预先初始化Predictor，在用户上传图片时仅调用已经构建好的Predictor的`predict()`方法，得到结果。

你好，我采用类似的方法，已经实现了将它封装成一个类，只需要加载模型一次，后面就只调用输出部分即可，但是刚刚发现了一个问题：  我使用同一张图像，循环推理100次，发现只有前两次有分割结果，后面都为空白，请问这是什么问题呢，以及存在的解决办法？

环境：jetpack 4.5
部属方法：c++ inferrence  gpu tensorrt

期待您的答复，感谢～ 
![xxxxxx屏幕截图](https://user-images.githubusercontent.com/59963025/197101120-1f211b6b-e137-46c6-90c6-4aac2401d556.png)

**Comment by Bobholamovic** [comment]
你好，我们这边没有观测到这种情况。请问方便提供一下核心代码，或者描述一下推理和存图的逻辑么？

**Comment by di-dii** [comment]
> 

谢谢你们的工作。

昨天检查了一下，是我自己的问题，把输入图像原地操作了（归一化），所以后面再输入这个图像就得不到正常的检测结果了。
